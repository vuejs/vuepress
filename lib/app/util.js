export function pathToComponentName (path) {
  if (path.charAt(path.length - 1) === '/') {
    return `page${path.replace(/\//g, '-') + 'index'}`
  } else {
    return `page${path.replace(/\//g, '-').replace(/\.html$/, '')}`
  }
}

export function findPageForPath (pages, path) {
  for (let i = 0; i < pages.length; i++) {
    const page = pages[i]
    if (page.path === path) {
      return page
    }
  }
}
