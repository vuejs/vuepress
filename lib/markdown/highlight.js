const chalk = require('chalk')
const prism = require('prismjs')
const loadLanguages = require('prismjs/components/index')
const escapeHtml = require('escape-html')
const logger = require('../util/logger')

// required to make embedded highlighting work...
loadLanguages(['markup', 'css', 'javascript'])

function wrap (code, lang) {
  if (lang === 'text') {
    code = escapeHtml(code)
  }
  return `<pre v-pre class="language-${lang}"><code>${code}</code></pre>`
}

module.exports = (str, lang) => {
  if (!lang) {
    return wrap(str, 'text')
  }
  lang = lang.toLowerCase()
  const rawLang = lang
  if (lang === 'vue' || lang === 'html') {
    lang = 'markup'
  }
  if (lang === 'md') {
    lang = 'markdown'
  }
  if (lang === 'ts') {
    lang = 'typescript'
  }
  if (lang === 'py') {
    lang = 'python'
  }
  if (!prism.languages[lang]) {
    try {
      loadLanguages([lang])
    } catch (e) {
      logger.warn(chalk.yellow(`[vuepress] Syntax highlight for language "${lang}" is not supported.`))
    }
  }
  if (prism.languages[lang]) {
    const code = prism.highlight(str, prism.languages[lang], lang)
    return wrap(code, rawLang)
  }
  return wrap(str, 'text')
}
