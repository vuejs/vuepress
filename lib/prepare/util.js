const path = require('path')
const spawn = require('cross-spawn')
const fs = require('fs-extra')
const globby = require('globby')

const tempPath = path.resolve(__dirname, '../app/.temp')
fs.ensureDirSync(tempPath)

const tempCache = new Map()
exports.writeTemp = async function (file, content) {
  // cache write to avoid hitting the dist if it didn't change
  const cached = tempCache.get(file)
  if (cached !== content) {
    await fs.writeFile(path.join(tempPath, file), content)
    tempCache.set(file, content)
  }
}

exports.writeEnhanceTemp = async function (destName, srcPath) {
  await exports.writeTemp(
    destName,
    fs.existsSync(srcPath)
      ? `export { default } from ${JSON.stringify(srcPath)}`
      : `export default function () {}`
  )
}

const indexRE = /(^|.*\/)(index|readme)\.md$/i
const extRE = /\.(vue|md)$/

exports.fileToPath = function (file) {
  if (exports.isIndexFile(file)) {
    // README.md -> /
    // foo/README.md -> /foo/
    return file.replace(indexRE, '/$1')
  } else {
    // foo.md -> /foo.html
    // foo/bar.md -> /foo/bar.html
    return `/${file.replace(extRE, '').replace(/\\/g, '/')}.html`
  }
}

exports.fileToComponentName = function (file) {
  let normalizedName = file
    .replace(/\/|\\/g, '-')
    .replace(extRE, '')
  if (exports.isIndexFile(file)) {
    normalizedName = normalizedName.replace(/readme$/i, 'index')
  }
  const pagePrefix = /\.md$/.test(file) ? `page-` : ``
  return `${pagePrefix}${normalizedName}`
}

exports.isIndexFile = function (file) {
  return indexRE.test(file)
}

exports.resolveComponents = async function (sourceDir) {
  const componentDir = path.resolve(sourceDir, '.vuepress/components')
  if (!fs.existsSync(componentDir)) {
    return
  }
  return exports.sort(await globby(['**/*.vue'], { cwd: componentDir }))
}

exports.sort = function (arr) {
  return arr.sort((a, b) => {
    if (a < b) return -1
    if (a > b) return 1
    return 0
  })
}

exports.encodePath = function (userpath) {
  return userpath.split('/').map(item => encodeURIComponent(item)).join('/')
}

exports.getGitLastUpdatedTimeStamp = function (filepath) {
  return parseInt(spawn.sync('git', ['log', '-1', '--format=%ct', filepath]).stdout.toString('utf-8')) * 1000
}
