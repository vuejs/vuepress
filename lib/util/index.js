const spawn = require('cross-spawn')
const parseHeaders = require('./parseHeaders')

exports.normalizeHeadTag = tag => {
  if (typeof tag === 'string') {
    tag = [tag]
  }
  const tagName = tag[0]
  return {
    tagName,
    attributes: tag[1] || {},
    innerHTML: tag[2] || '',
    closeTag: !(tagName === 'meta' || tagName === 'link')
  }
}

exports.applyUserWebpackConfig = function (userConfig, config, isServer) {
  const merge = require('webpack-merge')
  if (typeof userConfig === 'object') {
    return merge(config, userConfig)
  }
  if (typeof userConfig === 'function') {
    const res = userConfig(config, isServer)
    if (res && typeof res === 'object') {
      return merge(config, res)
    }
  }
  return config
}

exports.inferTitle = function (frontmatter) {
  if (frontmatter.data.home) {
    return 'Home'
  }
  if (frontmatter.data.title) {
    return parseHeaders(frontmatter.data.title)
  }
  const match = frontmatter.content.trim().match(/^#+\s+(.*)/)
  if (match) {
    return parseHeaders(match[1])
  }
}

exports.parseFrontmatter = content => {
  const matter = require('gray-matter')
  const toml = require('toml')

  return matter(content, {
    excerpt_separator: '<!-- more -->',
    engines: {
      toml: toml.parse.bind(toml),
      excerpt: false
    }
  })
}

const LRU = require('lru-cache')
const cache = LRU({ max: 1000 })

exports.extractHeaders = (content, include = [], md) => {
  const key = content + include.join(',')
  const hit = cache.get(key)
  if (hit) {
    return hit
  }

  const tokens = md.parse(content, {})

  const res = []
  tokens.forEach((t, i) => {
    if (t.type === 'heading_open' && include.includes(t.tag)) {
      const title = parseHeaders(tokens[i + 1].content)
      const slug = t.attrs.find(([name]) => name === 'id')[1]
      res.push({
        level: parseInt(t.tag.slice(1), 10),
        title,
        slug: slug || md.slugify(title)
      })
    }
  })

  cache.set(key, res)
  return res
}

exports.getGitLastModifiedTimeStamp = filepath => {
  return parseInt(spawn.sync('git', ['log', '-1', '--format=%ct', filepath]).stdout.toString('utf-8'))
}
