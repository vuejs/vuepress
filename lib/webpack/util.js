const { readFile } = require('fs-extra')
const { resolve, dirname } = require('path')

const asyncReplace = async (str, regex, aReplacer) => {
  regex = new RegExp(regex, 'g')
  const replacedParts = []
  let match
  let i = 0
  while ((match = regex.exec(str)) !== null) {
    // put non matching string
    replacedParts.push(str.slice(i, match.index))
    // call the async replacer function with the matched array spreaded
    replacedParts.push(aReplacer(...match))
    i = regex.lastIndex
  }

  // put the rest of str
  replacedParts.push(str.slice(i))

  // wait for aReplacer calls to finish and join them back into string
  return (await Promise.all(replacedParts)).join('')
}

const mdExplodeIncludes = exports.mdExplodeIncludes = async ({ dir, src }) => {
  const deps = []

  return {
    explodedSrc: await asyncReplace(src, /<!--\s*include\s+([^\s]+)\s*-->/g, async (match, path) => {
      try {
        const absolutePath = resolve(dir, path)
        const content = await readFile(absolutePath, 'utf8')

        // recursively explode the included file
        const { explodedSrc, dependencies } = await mdExplodeIncludes({
          dir: dirname(absolutePath),
          src: content
        })

        deps.push(absolutePath, ...dependencies)
        return explodedSrc
      } catch (e) {
        return match
      }
    }),
    dependencies: deps
  }
}
