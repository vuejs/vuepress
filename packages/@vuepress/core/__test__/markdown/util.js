import fs from 'fs-extra'
import LRU from 'lru-cache'
import path from 'path'

const cache = LRU({ max: 1000 })

export function Md () {
  return require('markdown-it')()
}

export async function getFragment (name) {
  let content = cache.get(name)
  if (content) {
    return content
  }
  const target = path.resolve(__dirname, `fragments/${name}.md`)
  content = await fs.readFile(target, 'utf-8')
  cache.set(name, content)
  return content
}
