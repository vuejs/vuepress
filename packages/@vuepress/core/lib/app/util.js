import Vue from 'vue'
import layoutComponents from '@internal/layout-components'
import pageComponents from '@internal/page-components'

const asyncComponents = Object.assign({}, layoutComponents, pageComponents)

export function isPageExists (pageKey) {
  return Boolean(pageComponents[pageKey])
}

export function isPageLoaded (pageKey) {
  return Boolean(Vue.component(pageKey))
}

export function getPageAsyncComponent (pageKey) {
  return pageComponents[pageKey]
}

export function isLayoutExists (layout) {
  return Boolean(layoutComponents[layout])
}

export function isLayoutLoaded (layout) {
  return Boolean(Vue.component(layout))
}

export function getLayoutAsyncComponent (pageKey) {
  return layoutComponents[pageKey]
}

export function ensureAsyncComponentsLoaded (...names) {
  return Promise.all(names.filter(v => v).map(async (name) => {
    if (!Vue.component(name) && asyncComponents[name]) {
      const comp = await asyncComponents[name]()
      Vue.component(name, comp.default)
    }
  }))
}

/**
 * Inject option to Vue SFC
 * @param {object} options
 * @param {string} key
 * @param {any} value
 */
export function injectComponentOption (options, key, value) {
  const arrayInject = () => {
    if (!options[key]) options[key] = []
    options[key].push(...value)
  }
  const objectInject = () => {
    if (!options[key]) options[key] = {}
    Object.assign(options[key], value)
  }
  // const primitiveInject = () => options[key] = value

  switch (key) {
  case 'components': objectInject(); break
  case 'mixins': arrayInject(); break
  default: throw new Error('Unknown option name.')
  }
}

export function findPageForPath (pages, path) {
  for (let i = 0; i < pages.length; i++) {
    const page = pages[i]
    if (page.path === path) {
      return page
    }
  }
  return {
    path: '',
    frontmatter: {}
  }
}

export function findPageByKey (pages, key) {
  for (let i = 0; i < pages.length; i++) {
    const page = pages[i]
    if (page.key === key) {
      return page
    }
  }
  return {
    path: '',
    frontmatter: {}
  }
}

/**
 * Normalize config.
 * This utility is mainly for plugin developers. For some
 * plugins that need internationalize the text. but it's
 * not recommenbded to let plugin care about to the internal
 * i18n implementation, so this utility was born.
 *
 *
 * Usage:
 *
 * import { normalizeConfig } from '@app/util'
 * export default {
 *   data () {
 *     return { config }
 *   }
 *   computed: {
 *     normalizedConfig() {
 *       return normalizeConfig(this, config)
 *     }
 *   }
 * }
 *
 *
 * e.g.
 *
 * Config: : 'Text'
 * Normalized Config: 'Text'
 *
 * Config: : { '/': 'Text', '/zh/': '文本' }
 * Normalized Config: 'Text' or '文本'
 *
 * @param {Vue} component
 * @param {any} rawConfig
 * @returns {any}
 */
export function normalizeConfig (component, rawConfig) {
  const { $localePath } = component
  if (typeof rawConfig === 'object' && rawConfig[$localePath]) {
    return rawConfig[$localePath]
  }
  return rawConfig
}
