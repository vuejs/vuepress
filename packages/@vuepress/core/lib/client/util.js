import Vue from 'vue'
import layoutComponents from '@internal/layout-components'
import pageComponents from '@internal/page-components'

/**
 * Create a cached version of a pure function.
 */
function cached (fn) {
  const cache = Object.create(null)
  // eslint-disable-next-line func-names
  return function cachedFn (str) {
    const hit = cache[str]
    // eslint-disable-next-line no-return-assign
    return hit || (cache[str] = fn(str))
  }
}

/**
 * Camelize a hyphen-delimited string.
 */
const camelizeRE = /-(\w)/g
const camelize = cached(str => {
  return str.replace(camelizeRE, (_, c) => c ? c.toUpperCase() : '')
})

/**
 * Hyphenate a camelCase string.
 */
const hyphenateRE = /\B([A-Z])/g
const hyphenate = cached(str => {
  return str.replace(hyphenateRE, '-$1').toLowerCase()
})

/**
 * Capitalize a string.
 */
const capitalize = cached(str => {
  return str.charAt(0).toUpperCase() + str.slice(1)
})

/**
 * This method was for securely getting Vue component when components
 * are named in different style.
 *
 * e.g. a component named `a-b` can be also getted by `AB`, It's the
 * same the other way round
 *
 * @param {function} getter a function of getting component by name
 * @param {string} name component's name
 * @returns {Component|AsyncComponent}
 */
export function getComponent (getter, name) {
  if (!name) return
  if (getter(name)) return getter(name)

  const isKebabCase = name.includes('-')
  if (isKebabCase) return getter(capitalize(camelize(name)))

  return getter(capitalize(name)) || getter(hyphenate(name))
}

const asyncComponents = Object.assign({}, layoutComponents, pageComponents)
const asyncComponentsGetter = name => asyncComponents[name]
const pageComponentsGetter = layout => pageComponents[layout]
const layoutComponentsGetter = layout => layoutComponents[layout]
const globalComponentsGetter = name => Vue.component(name)

export function getPageAsyncComponent (pageKey) {
  return getComponent(pageComponentsGetter, pageKey)
}

export function getLayoutAsyncComponent (layout) {
  return getComponent(layoutComponentsGetter, layout)
}

export function getAsyncComponent (name) {
  return getComponent(asyncComponentsGetter, name)
}

export function getVueComponent (name) {
  return getComponent(globalComponentsGetter, name)
}

export function ensureAsyncComponentsLoaded (...names) {
  return Promise.all(names.filter(v => v).map(async (name) => {
    if (!getVueComponent(name) && getAsyncComponent(name)) {
      const comp = await getAsyncComponent(name)()
      Vue.component(name, comp.default)
    }
  }))
}

/**
 * Inject option to Vue SFC
 * @param {object} options
 * @param {string} key
 * @param {any} value
 */
export function injectComponentOption (options, key, value) {
  const arrayInject = () => {
    if (!options[key]) options[key] = []
    options[key].push(...value)
  }
  const objectInject = () => {
    if (!options[key]) options[key] = {}
    Object.assign(options[key], value)
  }
  // const primitiveInject = () => options[key] = value

  switch (key) {
  case 'components': objectInject(); break
  case 'mixins': arrayInject(); break
  default: throw new Error('Unknown option name.')
  }
}

export function findPageForPath (pages, path) {
  for (let i = 0; i < pages.length; i++) {
    const page = pages[i]
    if (page.path === path) {
      return page
    }
  }
  return {
    path: '',
    frontmatter: {}
  }
}

export function findPageByKey (pages, key) {
  for (let i = 0; i < pages.length; i++) {
    const page = pages[i]
    if (page.key === key) {
      return page
    }
  }
  return {
    path: '',
    frontmatter: {}
  }
}

/**
 * Normalize config.
 * This utility is mainly for plugin developers. For some
 * plugins that need internationalize the text. but it's
 * not recommenbded to let plugin care about to the internal
 * i18n implementation, so this utility was born.
 *
 *
 * Usage:
 *
 * import { normalizeConfig } from '@app/util'
 * export default {
 *   data () {
 *     return { config }
 *   }
 *   computed: {
 *     normalizedConfig() {
 *       return normalizeConfig(this, config)
 *     }
 *   }
 * }
 *
 *
 * e.g.
 *
 * Config: : 'Text'
 * Normalized Config: 'Text'
 *
 * Config: : { '/': 'Text', '/zh/': '文本' }
 * Normalized Config: 'Text' or '文本'
 *
 * @param {Vue} component
 * @param {any} rawConfig
 * @returns {any}
 */
export function normalizeConfig (component, rawConfig) {
  const { $localePath } = component
  if (typeof rawConfig === 'object' && rawConfig[$localePath]) {
    return rawConfig[$localePath]
  }
  return rawConfig
}

/**
 * Set global info in `window.__VUEPRESS__` for debugging.
 *
 * @param {string}key
 * @param {any} value
 */
export function setGlobalInfo (key, value) {
  if (typeof window === 'undefined' || !window.__VUEPRESS__) {
    return
  }
  window.__VUEPRESS__[key] = value
}
