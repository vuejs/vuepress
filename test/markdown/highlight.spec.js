import { Md, getFragment } from './util'
import highlight from '@/markdown/highlight.js'

const md = Md()
const mdH = Md().set({ highlight })

describe('highlight', () => {
  test('should highlight code', async () => {
    const input = await getFragment('code')
    const output1 = md.render(input)
    const output2 = mdH.render(input)
    expect(output1 === output2).toBe(false)
    expect(output2).toMatchSnapshot()
  })
})
