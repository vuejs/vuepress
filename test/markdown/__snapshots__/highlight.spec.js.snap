// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`highlight should highlight code 1`] = `
<pre v-pre class="language-js"><code><span class="token keyword">new</span> <span class="token class-name">Vue</span><span class="token punctuation">(</span><span class="token punctuation">)</span>
</code></pre>
`;
